File: lib/repositories/recurring_obligation.rb
module Repositories
  RecurringObligation = Struct.new(:recurring_obligation_id, :envelope, :description, :every, :period, :start_on, :end_on, :amount, :created_at, :updated_at) do
    alias_method :id, :recurring_obligation_id

    def envelope_name
      envelope.name
    end
  end
end

File: src/main/scala/almoneya/automation/RecurringObligation.scala
package almoneya.automation

import almoneya._
import org.joda.time.LocalDate

case class RecurringObligation(priority: Priority,
                               name: ObligationName,
                               target: Amount,
                               balance: Amount,
                               dueOn: LocalDate,
                               period: Period,
                               every: Every,
                               endOn: Option[LocalDate] = None) extends FundingGoal with DueOnOrAfter {
    override def payoutsOnOrBefore(cutoffOn: LocalDate): Seq[LocalDate] =
        payoutEventsStream.takeWhile(_.compareTo(cutoffOn) <= 0)

    override def startOn: LocalDate = dueOn
}
